### PR TITLE
feat(DEVOPS-1803): Publish joy-generator image and chart

### PR DESCRIPTION
**Pull request by alexstojda**
- feat(DEVOPS-1803): Publish joy-generator image to GCR
- feat(DEVOPS-1803): Create helm chart
- feat(DEVOPS-1803): Support using sealed secrets
- feat(DEVOPS-1803): Add shared api token authorization as expected by ArgoCD
- feat(DEVOPS-1803): Publish chart

### CODE REVIEW

**Review by silphid — CHANGES_REQUESTED** [review]
Wow, quite a piece of work! Great job! 🚀

File: .github/workflows/build-publish.yaml [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        uses: docker/setup-buildx-action@v2
+
+      - uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+
+      - name: Log in to the Container registry
+        uses: docker/login-action@v2.1.0
+        with:
+          registry: ${{ env.REGISTRY }}
```
[the remark this answers is not in the document]
Comment (alexstojda) [reply .github/workflows/build-publish.yaml:69]: Copy-paste error! Should be hardcoded, I copied parts of this workflow from a personal project.

File: .github/workflows/build-publish.yaml [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    runs-on: ubuntu-latest
+    steps:
+      - name: Setup Docker Buildx
+        uses: docker/setup-buildx-action@v2
+      - uses: actions/checkout@v3
+      - name: Build Docker image
+        uses: docker/build-push-action@v3.2.0
+        with:
+          context: .
+          push: false
```
[the remark this answers is not in the document]
Comment (alexstojda) [reply .github/workflows/build-publish.yaml:45]: Publish has a bunch of additional steps that don't need to be run during a PR build, so separating them makes for a cleaner build pipeline, IMO. And since we are using the GitHub actions layer cache, there is no major sacrifice in performance.

File: .github/workflows/build-publish.yaml [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+      - name: Log in to the Container registry
+        uses: docker/login-action@v2.1.0
+        with:
+          registry: ${{ env.REGISTRY }}
+          username: ${{ github.actor }}
+          password: ${{ github.token }}
+
+      - uses: docker/setup-buildx-action@v2
+      - name: Log in to the Container registry
```
[the remark this answers is not in the document]
Comment (alexstojda) [reply .github/workflows/build-publish.yaml:74]: Oops! I'll fix that

File: .github/workflows/build-publish.yaml [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        if: ${{ steps.changelog.outputs.skipped == 'false' }}
+        uses: mikefarah/yq@v4
+        with:
+          cmd: yq -i '.appVersion = "${{ steps.changelog.outputs.version }}"' 'chart/Chart.yaml'
+
+      - name: Commit release
+        if: ${{ steps.changelog.outputs.skipped == 'false' }}
+        run: |
+          git config user.name "GitHub Actions"
+          git config user.email "41898282+github-actions[bot]@users.noreply.github.com"
```
[the remark this answers is not in the document]
Comment (alexstojda) [reply .github/workflows/build-publish.yaml:100]: After LOTS of searching, this seems to be the email to use in order to get the GitHub actions username icon on the commit message haha! 

image

Source: https://github.com/fregante/setup-git-user

File: .github/workflows/build-publish.yaml [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+          fallback-version: 0.0.0
+          skip-commit: true
+          skip-version-file: true
+          git-push: false
+
+      - name: Update appVersion in chart
+        if: ${{ steps.changelog.outputs.skipped == 'false' }}
+        uses: mikefarah/yq@v4
+        with:
+          cmd: yq -i '.appVersion = "${{ steps.changelog.outputs.version }}"' 'chart/Chart.yaml'
```
[the remark this answers is not in the document]
Comment (alexstojda) [reply .github/workflows/build-publish.yaml:94]: Yep! I plan to make that change when I add the chart publish step(s)

File: chart/templates/_helpers.tpl [hunk 6]
```diff
@@ -0,0 +1,51 @@
+{{/*
+Expand the name of the chart.
+*/}}
+{{- define "joy-generator.name" -}}
```
[the remark this answers is not in the document]
Comment (alexstojda) [reply chart/templates/_helpers.tpl:4]: possibly yes, but this is actually standard helm stuff. This `_helpers.tpl` is generated by default when you do `helm create`

File: chart/templates/hpa.yaml
[the remark this answers is not in the document]
Comment (alexstojda) [reply chart/templates/hpa.yaml:1]: Again this is from the boilerplate chart created with `helm create` 

Honestly, I don't know if we'll need it, but I figured I'd keep it just in case, since its already setup :)

File: .github/workflows/build-publish.yaml [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+      - name: Extract metadata (tags, labels) for Docker
+        id: meta
+        if: ${{ steps.changelog.outputs.skipped == 'false' }}
+        uses: docker/metadata-action@v4.1.1
+        with:
+          images: ghcr.io/${{ github.repository }}
+          tags: |
+            type=raw,value=latest
+            type=semver,pattern={{version}},value=${{ steps.changelog.outputs.tag }}
```
[the remark this answers is not in the document]
Comment (alexstojda) [reply .github/workflows/build-publish.yaml:112]: This is the magic of the docker-metadata action! :D 

See [here](https://github.com/docker/metadata-action#typesemver)

TL;DR using `type=semver` you pass the semver-compliant git tag and a pattern, and it will create the appropriate docker tag for the image.

File: .github/workflows/build-publish.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (silphid) [reply .github/workflows/build-publish.yaml:94]: Why not use that convenience action then?
[fregante/setup-git-user](https://github.com/fregante/setup-git-user)

File: .github/workflows/build-publish.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (silphid) [reply .github/workflows/build-publish.yaml:106]: That I understand, but my point was that `.tag` is _not_ semver-compliant, right?

File: .github/workflows/build-publish.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (alexstojda) [reply .github/workflows/build-publish.yaml:94]: Tbh I didn't see the need to use the action, it's an unnecessary 3rd party dependency for a simple 2 lines of shell :P

File: .github/workflows/build-publish.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (alexstojda) [reply .github/workflows/build-publish.yaml:106]: Oh, yeah, I see your point. I just went with this way since its expected to use the tag, according to the docs 🤷🏼

[the remark this answers is not in the document]
Comment (silphid) [reply .github/workflows/build-publish.yaml:106]: But that would probably break if tag prefix was anything else than `""`, right?

[the remark this answers is not in the document]
Comment (silphid) [reply .github/workflows/build-publish.yaml:106]: So we'll use `.version`, right?

[the remark this answers is not in the document]
Comment (alexstojda) [reply .github/workflows/build-publish.yaml:106]: > But that would probably break if tag prefix was anything else than "", right?

Probably, but I honestly don't know. The docs don't say much about that. 

> So we'll use .version, right?

Sure, if that helps you sleep at night ;) The version tagged on this project will be the same for both the chart and app, `v0.0.0`, so it doesn't really make a difference

File: .github/workflows/build-publish.yaml [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+          draft: false
+          name: ${{ steps.changelog.outputs.tag }}
+          tag: ${{ steps.changelog.outputs.tag }}
+          body: ${{ steps.changelog.outputs.clean_changelog }}
+          token: ${{ github.token }}
+          artifacts: ${{ steps.chart-release.outputs.artifact-path }}
+
+      - name: Deploy to GitHub Pages
+        if: ${{ steps.changelog.outputs.skipped == 'false' }}
+        id: deployment
```
Comment (silphid) [inline .github/workflows/build-publish.yaml:174]: Not needed
```suggestion
```

**Review by silphid — APPROVED** [review]
LGTM!
